### PR TITLE
Payments Framework: TEM-56, TEM-65, TEM-66

### PR DESCRIPTION
**Pull request by postables**
## :construction_worker: Purpose
Lays the foundation for TEMPORAL's payment system.
Solves TEM-56, TEM-65, TEM-66

The corresponding "service" is in a [different repository](https://github.com/RTradeLtd/Temporal_Payment-ETH), due to the tremendous licensing headache go-ethereum has and so that Temporal (the core code base at least) won't have to deal with that. It's nowhere near finished, however it also lays the foundation for building our the "backend" to our payment system. As of right now, I'm not sure if that service will be strictly limited to Ethereum, or if it will include other supported cryptocurrencies. This will largely depend on the licensing requirements of packages we will end up using.

"Technically" the following payment methods are supported (but the beta is free! so it wont see much use other than testing for awhile):
* [x] ETH
* [x] RTC
* [ ] XMR
* [ ] DASH
* [ ] BTC
* [ ] LTC

Up next, I will be integrating DASH and XMR payments.

## :rocket: Changes
Add payment gates to a select number of routes:
* IPFS File Uploads and Pinning
* IPNS Entry Creation
* IPFS Key Creation
* IPFS Cluster Pinning
* IPFS PubSub

Users are now able to purchase credits by fetching a deposit address, sending crypto to it, and then providing us with the transaction information. For now we are using a single deposit address per payment method, as opposed to unique deposit addresses per user. Having unique deposit addresses would make the process much simpler, but it also introduces the requirement for us to handle private keys, exposing us as a greater target for attackers. Currently unique deposit addresses are pretty low on the priority list. Once the larger payment system as a whole is finalized and fleshed out, we will integrate unique deposti addresses.

## :warning: Breaking Changes
Yes (well, for the beta at least). When this change goes live on the staging servers, we will need to automatically grant credits to users on sign up.

### CODE REVIEW

File: api/utils.go [hunk 1]
```diff
@@ -59,3 +60,56 @@ func (api *API) FileSizeCheck(size int64) error {
 	}
 	return nil
 }
+
+func (api *API) getDepositAddress(paymentType string) (string, error) {
+	switch paymentType {
+	case "eth", "rtc":
+		return "0xc7459562777DDf3A1A7afefBE515E8479Bd3FDBD", nil
+	case "btc":
+		return "0", nil
```
Comment (bobheadxi) [inline api/utils.go]: why not empty string, which is perhaps a bit more of a standard 0-value for strings? or is there another purpose?

File: models/user.go [hunk 2]
```diff
@@ -59,6 +62,8 @@ func (um *UserManager) CheckIfUserHasAccessToNetwork(username, networkName strin
 	}
 	return false, nil
 }
+
+// AddIPFSNetworkForUser is used to update a users allowed private ipfs networks
```
Comment (bobheadxi) [inline models/user.go]: docs 😍😍

File: api/routes_rtfs.go [hunk 3]
```diff
@@ -60,24 +60,38 @@ func (api *API) pinHashLocally(c *gin.Context) {
 		Fail(c, err)
 		return
 	}
-
+	shell, err := rtfs.Initialize("", "")
```
[the remark this answers is not in the document]
Comment (postables) [reply api/routes_rtfs.go]: This is a good point, it's "deprecated" syntax from previous temporal versions (the first param, pubsub topic), mixed in with host to connect to (the second param). I was going to open up a ticket shortly for me to revamp the initialization of the rtfs package.

File: api/routes_payment.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+	case "eth":
+		return utils.RetrieveUsdPrice("ethereum")
+	case "xmr":
+		return utils.RetrieveUsdPrice("monero")
+	case "btc":
+		return utils.RetrieveUsdPrice("bitcoin")
+	case "ltc":
+		return utils.RetrieveUsdPrice("litecoin")
+	case "rtc":
+		return 0.125, nil
```
[the remark this answers is not in the document]
Comment (postables) [reply api/routes_payment.go]: Yes, I will add as a constant.

File: api/routes_rtfs.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (bobheadxi) [reply api/routes_rtfs.go]: sounds good!

**Pull request merged**